src/cli/traceViewer/web/ui/sourceTab.tsx

```diff
@@ -21,27 +21,82 @@ import './sourceTab.css';
 import '../../../../third_party/highlightjs/highlightjs/tomorrow.css';
 import * as highlightjs from '../../../../third_party/highlightjs/highlightjs';
 
+type StackInfo = string | {
+  frames: {
+    filePath: string,
+    fileName: string,
+    lineNumber: number,
+    functionName: string,
+  }[];
+  fileContent: Map<string, string>;
+};
+
 export const SourceTab: React.FunctionComponent<{
   actionEntry: ActionEntry | undefined,
 }> = ({ actionEntry }) => {
-  const location = React.useMemo<{ fileName?: string, lineNumber?: number, value?: string }>(() => {
+  const [lastAction, setLastAction] = React.useState<ActionEntry | undefined>();
+  const [selectedFrame, setSelectedFrame] = React.useState<number>(0);
+  const [needReveal, setNeedReveal] = React.useState<boolean>(false);
```

Comment:
note/nit: React internally infers the types automatically if you pass them into useState. For cases to allow e.g. null you need to use their generic.